Dockerfile

```diff
@@ -1 +1,41 @@
+ARG ARCH="amd64"
+ARG OS="linux"
+FROM 345280441424.dkr.ecr.ap-south-1.amazonaws.com/ark_base:latest
+LABEL maintainer="Armedia, LLC"
 
+ARG ARCH="amd64"
+ARG OS="linux"
+ARG VER="2.28.1"
+ARG PKG="prometheus"
+ARG SRC="${PKG}-${VER}.${OS}-${ARCH}"
+
+RUN curl \
+		-L "https://github.com/prometheus/${PKG}/releases/download/v${VER}/${SRC}.tar.gz" \
+		-o package.tar.gz && \
+	tar -xzvf "package.tar.gz" && \
+	mkdir -pv "/prometheus" && \
+	mkdir -pv "/etc/prometheus" && \
+	mkdir -pv "/usr/share/prometheus" && \
+	mv -vif "${SRC}/LICENSE"                     "/LICENSE" && \
+	mv -vif "${SRC}/NOTICE"                      "/NOTICE" && \
+	mv -vif "${SRC}/prometheus"                  "/bin/prometheus" && \
+	mv -vif "${SRC}/promtool"                    "/bin/promtool" && \
+	mv -vif "${SRC}/prometheus.yml"              "/etc/prometheus/prometheus.yml" && \
+	mv -vif "${SRC}/console_libraries/"          "/usr/share/prometheus/" && \
+	mv -vif "${SRC}/consoles/"                   "/usr/share/prometheus/" && \
+	ln -s "/usr/share/prometheus/console_libraries" "/etc/prometheus" && \
+	ln -s "/usr/share/prometheus/consoles"          "/etc/prometheus" && \
+	chown -R nobody:nobody "/etc/prometheus" "/prometheus" && \
+	rm -rvf "${SRC}" package.tar.gz
+
+#COPY npm_licenses.tar.bz2                        /npm_licenses.tar.bz2
+
+USER       nobody
```

Comment:
Please create a user using `useradd` rather than using `nobody` (which is now promoted to somebody)

Reply:
The user "nobody" is part of the base CentOS 7 install, and is meant precisely for this sort of thing. Are you saying we should actually have a specific "prometheus" user (in this case)?

Reply:
My understanding of Linux is that the user `nobody` is used to isolate things. Here Prometheus will run as `nobody` and thus will have read/write access to `/etc/prometheus` and `/prometheus`, which is bad. Prometheus should have read access only to what it needs to read, and write access only to what it needs to write (which I suppose is going to be only its database).

Reply:
Well, /etc/prometheus and /prometheus (being changed to /app/conf and /app/data respectively) are the only areas the process should read and write from explicitly.

I was seeking to avoid having to create users and start "littering" the exported volumes with UIDs that won't exist outside. Furthermore, if each container creates its own user there's a chance all the UIDs will be the same (b/c they're the first user created in the OS image), so that kind of defeats the purpose of isolation in the case of volumes...

Reply:
@drivera-armedia I am quite sure that Docker does not permit one container to see volumes that are not mounted to it, whether the UIDs are the same or not. Also, I don't think we should preoccupy ourselves with users on the host, which might be minikube, a docker-compose running on Windows, an EKS node (derived for Amazon Linux), etc.

You say that `/etc/prometheus` must be writeable by Prometheus? That sounds very odd to me... Could you please elaborate?

`/prometheus` is where Prometheus will store its database, is that correct?

I think creating a dedicated user is more explicit, and also allows you more control over this user (eg: location of the home directory, which shell to run, whether you want to disallow shell access entirely for this user, etc.)

Reply:
If you're concerned about denying write-access, one can always mount volumes in read-only mode, regardless of the actual permissions granted at the O/S level. I think this is threading a bit thinly in the sense that we're solving a problem we've yet to have, and I don't think we have a pressing need to resolve now (not to mention that the aforementioned read-only volume mounting would solve it summarily with no change to permissions).

The Dockerfile has been modified to use /app/conf and /app/data, which should match what you have in other places.  The value of ensuring that the UID doesn't collide with a host O/S UID is to help minimize inadvertent modifications of the contained data: you want to change it? you'll need to make a concerted effort to change it ...

Reply:
@drivera-armedia What about denying write access to a file/directory that is not mounted?

The "problem" I am trying to solve is us adhering to best practices, one of the most important being the principle of least privilege.

Reply:
I get that. But as I mentioned elsewhere, unless we have perfect knowledge that the application will never attempt to write to a read-only volume, or that any such attempt will not result in a problem, I think it's a beautiful yet risky concept to pursue at this time.

Reply:
> I get that. But as I mentioned elsewhere, unless we have perfect knowledge that the application will never attempt to write to a read-only volume, or that any such attempt will not result in a problem, I think it's a beautiful yet risky concept to pursue at this time.

OK, I understand.

> Finally, you're never going to be able to deny access to a file/directory that's not mounted unless you resort to permissions trickery

I very much do not understand that statement... Could you please expand? Thanks

Reply:
All right, thanks